### PR TITLE
Add tox & pre-commit

### PR DESCRIPTION
**Pull request by Fireye04**

### CODE REVIEW

**Comment by timj** [comment]
I don't know what the SQuaRE protocol is for this and maybe the person you are working with can jump in, but I'd really like the black/isort etc configurations to live in `pyproject.toml` rather than in the pre-commit config file. The problem is that now if you run `black` in the directory you might get a different answer to what you would get if you ran pre-commit.

Something like:

```toml
[tool.black]
line-length = 110
target-version = ["py311"]

[tool.isort]
profile = "black"
line_length = 110
```
means that black and isort run locally will match the versions run by pre-commit.

(I don't know what the sqre line length convention is, pipelines use 110).

**Comment by jonathansick** [comment]
Hey @Fireye04, you can use the SQuaRE PyPI package template to guide you here. See https://github.com/lsst/templates/tree/main/project_templates/square_pypi_package/example as an example project. Essentially setup.py gets replaced with pyproject.toml. SQuaRE's configuration for black and isort is in that pyproject.toml (note it's a bit different from DM in general). You can also see/copy our basic `.pre-commit-config.yaml` file there.

**Comment by Fireye04** [comment]
Awesome, thanks for the tip. I ended up copying a lot of this stuff from the tox documentation and the phalanx repo that I happened to have a local clone of, so having a good template sounds like a great idea lol. I'll get on it :)

**Comment by jonathansick** [comment]
Next, try adding a `.github/workflows` directory. It should contain a `ci.yaml` workflow that looks like this: https://github.com/lsst/templates/blob/main/project_templates/square_pypi_package/example/.github/workflows/ci.yaml

That workflow has a bunch of jobs, and you may need to delete the ones that aren't relevant. Probably keep `lint`, `tests`, `test-packages`, `pypi` (if this gets published to PyPI).

For the `test` job, look at the tox environments list; you'll probably need to drop the `typing` environment.

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0 # full history for setuptools_scm
+
+      - name: Build and publish
+        uses: lsst-sqre/build-and-publish-to-pypi@v2
+        with:
+          python-version: "3.11"
```
Comment (jonathansick) [inline .github/workflows/ci.yaml]: On second look, it seems we're not pushing this to PyPI? In that case, you should delete this job.

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    "Programming Language :: Python",
+    "Programming Language :: Python :: 3",
+    "Programming Language :: Python :: 3.11",
+    "Intended Audience :: Developers",
+    "Natural Language :: English",
+    "Operating System :: POSIX",
+    "Typing :: Typed",
+]
+requires-python = ">=3.11"
+dependencies = []
```
Comment (jonathansick) [inline pyproject.toml]: It looks like you need to bring over the dependencies originally in `setup.py`:

```suggestion
dependencies = [
    "jira",
    "tabulate",
    "flask",
    "graphviz"
]
```

Comment (Fireye04) [reply pyproject.toml]: oh snap that makes sense! I was trying to do that in the wrong way entirely lmao

Comment (jonathansick) [reply pyproject.toml]: The GitHub preview seems to be indicating I got the toml syntax wrong; not sure?

**Review by jonathansick — APPROVED** [review]
From my perspective this looks great. There's a bunch more work that can be done, like turning everything in `bin/` into a script declared in pyproject.toml (https://setuptools.pypa.io/en/latest/userguide/entry_point.html#console-scripts), but this is a achieves the goal of having the package run under GitHub Actions with modern linting and testing.

File: pyproject.toml (hunk 2 above)
Comment (Fireye04) [reply pyproject.toml]: Yeah idk why it's highlighted red, but it looks fine to me. I looked up the toml docs and cross referenced another repo and it looks/ runs well.

**Comment by jonathansick** [comment]
After you merge this I can change the repo settings so that `master` becomes `main`.

**Comment by Fireye04** [comment]
Alright, awesome. I'll go ahead and merge it if there are no complaints then. I might start up a new PR for the pyproject bin scripts if frossie doesn't have anything better for me to do after finishing this one up. Thanks for the help :)

**Pull request merged**